File: python/binary-search/binary_search.py
def binary_search(list_of_numbers, number):
    lower_index, upper_index = 0, len(list_of_numbers) - 1
    while lower_index <= upper_index:
        middle_index = (lower_index + upper_index) // 2
        if list_of_numbers[middle_index] == number:
            return middle_index
        elif list_of_numbers[middle_index] < number:
            lower_index = middle_index + 1
        else:
            upper_index = middle_index - 1
    raise ValueError('Not found')

File: java/binary-search/src/main/java/BinarySearch.java
import java.util.List;

public class BinarySearch<T extends Comparable<? super T>> {
  List<T> elements;

  BinarySearch(List<T> elements) {
    this.elements = elements;
  }

  int indexOf(T target) throws ValueNotFoundException {
    int lowerIndex = 0;
    int upperIndex = elements.size() - 1;
    while (lowerIndex <= upperIndex) {
      int middleIndex = (lowerIndex + upperIndex) / 2;
      int cmp = elements.get(middleIndex).compareTo(target);
      if (cmp == 0) {
        return middleIndex;
      }

      if (cmp < 0) {
        lowerIndex = middleIndex + 1;
      } else {
        upperIndex = middleIndex - 1;
      }
    }

    throw new ValueNotFoundException("Value not in array");
  }
}
